c++/src/kj/compat/http-test.c++

```diff
@@ -5527,6 +5527,151 @@ KJ_TEST("HttpServer.listenHttp() doesn't prematurely terminate if an accepted co
   KJ_ASSERT(!promise.poll(waitScope));
 }
 
+KJ_TEST("HttpServer handles disconnected exception for clients disconnecting after headers") {
+  // This test case reproduces a race condition where a client could disconnect after the server
+  // sent response headers but before it sent the response body, resulting in a broken pipe
+  // "disconnected" exception when writing the body.  The default handler for application errors
+  // tells the server to ignore "disconnected" exceptions and close the connection, but code
+  // after the handler exercised the broken connection, causing the server loop to instead fail
+  // with a "failed" exception.
+
+  KJ_HTTP_TEST_SETUP_IO;
+  kj::TimerImpl timer(kj::origin<kj::TimePoint>());
+  auto pipe = KJ_HTTP_TEST_CREATE_2PIPE;
+
+  class SendErrorHttpService final: public HttpService {
+    // HttpService that serves an error page via sendError().
+  public:
+    SendErrorHttpService(HttpHeaderTable& headerTable): headerTable(headerTable) {}
+    kj::Promise<void> request(
+        HttpMethod method, kj::StringPtr url, const HttpHeaders& headers,
+        kj::AsyncInputStream& requestBody, Response& responseSender) override {
+      return responseSender.sendError(404, "Not Found", headerTable);
+    }
+
+  private:
+    HttpHeaderTable& headerTable;
+  };
+
+  class DisconnectingAsyncIoStream final: public kj::AsyncIoStream {
+  public:
+    DisconnectingAsyncIoStream(AsyncIoStream& inner): inner(inner) {}
+
+    Promise<size_t> read(void* buffer, size_t minBytes, size_t maxBytes) override {
+      return inner.read(buffer, minBytes, maxBytes);
+    }
+    Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override {
+      return inner.tryRead(buffer, minBytes, maxBytes);
+    }
+
+    Maybe<uint64_t> tryGetLength() override { return inner.tryGetLength(); }
+
+    Promise<uint64_t> pumpTo(AsyncOutputStream& output, uint64_t amount) override {
+      return inner.pumpTo(output, amount);
+    }
+
+    Promise<void> write(const void* buffer, size_t size) override {
+      int writeId = writeCount++;
+      if (writeId == 0) {
+        // Allow first write (headers) to succeed.
+        auto promise = inner.write(buffer, size);
+        inner.shutdownWrite();
```

Comment:
I don't fully understand why, but it seemed that this call to `shutdownWrite()` was necessary for the race condition failure to manifest in `http-test`.  Without it, the failure only manifested when run as part of `http-socketpair-test`.  I assume there's some aspect of native pipes that makes it necessary.